### ISSUE TITLE
Phase 33.5 — ContinualOrchestrator: Unified continual learning pipeline with forgetting metrics

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#695]
## Overview

The `ContinualOrchestrator` is the capstone component of Phase 33, integrating the ElasticWeightConsolidator (33.1), ProgressiveNetExpander (33.2), ReplayMemoryManager (33.3), and CurriculumScheduler (33.4) into a unified continual learning pipeline. It provides strategy selection (regularization vs. replay vs. architectural), adaptive switching between methods based on task characteristics, and comprehensive forgetting metrics including Backward Transfer (BWT), Forward Transfer (FWT), and knowledge retention tracking.

## Key Responsibilities

- **Strategy selection** — Choose optimal combination of consolidation, expansion, and replay for each task
- **Adaptive method switching** — Monitor forgetting metrics and switch strategies if current approach degrades
- **Backward Transfer (BWT) tracking** — Measure accuracy changes on previous tasks after learning new ones
- **Forward Transfer (FWT) tracking** — Measure how previous knowledge accelerates new task learning
- **Knowledge retention dashboard** — Real-time metrics on per-task accuracy, capacity usage, replay effectiveness
- **Scenario detection** — Automatically classify learning scenario (task-incremental, domain-incremental, class-incremental per van de Ven & Tolias, 2019)
- **Checkpoint management** — Save/restore full continual learning state (model, buffers, Fisher matrices, column registry)
- **Pipeline configuration** — YAML-based configuration for all sub-components with sensible defaults

### Inputs
- `consolidator: ElasticWeightConsolidator` — From Phase 33.1
- `expander: ProgressiveNetExpander` — From Phase 33.2
- `replay_manager: ReplayMemoryManager` — From Phase 33.3
- `scheduler: CurriculumScheduler` — From Phase 33.4
- `strategy: Literal["regularization", "replay", "architectural", "hybrid", "auto"]`
- `config: ContinualConfig` — Full pipeline configuration
- `task_stream: AsyncIterator[Task]` — Incoming task stream

### Outputs
- `ContinualReport` — Per-task and aggregate metrics (BWT, FWT, average accuracy, forgetting rate)
- `RetentionMatrix` — T×T matrix of accuracy[task_i after learning task_j]
- `PipelineState` — Full serializable state for checkpoint/restore
- `StrategyLog` — Record of which strategy was used for each task and why

## Acceptance Criteria

- [ ] Hybrid strategy achieves highest average accuracy across all three CL scenarios
- [ ] Auto mode correctly selects best strategy >80% of the time
- [ ] BWT/FWT metrics computed accurately against ground truth evaluations
- [ ] Full pipeline state serializable and restorable within 10 seconds (100M param model)
- [ ] Handles task streams of 50+ tasks without memory leaks
- [ ] Configuration validated at startup with clear error messages
- [ ] End-to-end benchmarks on Split-MNIST, Split-CIFAR10, Permuted-MNIST
- [ ] Integration with Phase 32 multi-agent collaboration for distributed CL
- [ ] 95%+ test coverage, all public methods typed

## References

- Parisi et al. (2019) — Continual Lifelong Learning Survey
- van de Ven & Tolias (2019) — Three Scenarios for Continual Learning
- Schwarz et al. (2018) — Progress & Compress
- de Lange et al. (2021) — A Continual Learning Survey
- Planning: #690

**Comment by web3guru888** [comment]
## Implementation Notes — ContinualOrchestrator

### Python Skeleton

```python
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Dict, List, Literal, Optional
import json, time
import torch
from torch import Tensor, nn

@dataclass
class ContinualConfig:
    """Full pipeline configuration."""
    strategy: Literal["regularization", "replay", "architectural", "hybrid", "auto"] = "auto"
    ewc_lambda: float = 5000.0
    fisher_method: str = "diagonal"
    buffer_size: int = 5000
    replay_strategy: str = "prioritized"
    expansion_policy: str = "similarity_threshold"
    max_columns: int = 50
    ordering_strategy: str = "transfer_optimal"
    spaced_repetition: bool = True
    checkpoint_dir: str = "./checkpoints"

@dataclass
class ContinualReport:
    """Per-task and aggregate continual learning metrics."""
    task_id: str
    accuracy: float
    avg_accuracy: float                      # across all seen tasks
    backward_transfer: float                 # BWT: mean accuracy change on old tasks
    forward_transfer: float                  # FWT: zero-shot accuracy before training
    forgetting_rate: float                   # fraction of tasks with >5% degradation
    strategy_used: str
    elapsed_seconds: float

@dataclass
class RetentionMatrix:
    """T×T accuracy matrix: retention[i][j] = accuracy on task_i after training task_j."""
    task_ids: List[str]
    matrix: Tensor     # shape (T, T)

    def bwt(self) -> float:
        """Average backward transfer."""
        T = len(self.task_ids)
        if T < 2:
            return 0.0
        total = 0.0
        for i in range(T - 1):
            total += (self.matrix[i, T-1] - self.matrix[i, i]).item()
        return total / (T - 1)

    def fwt(self) -> float:
        """Average forward transfer (vs random init baseline)."""
        T = len(self.task_ids)
        if T < 2:
            return 0.0
        total = 0.0
        for i in range(1, T):
            total += self.matrix[i, i-1].item()  # zero-shot before training task_i
        return total / (T - 1)

@dataclass
class PipelineState:
    """Full serialisable state for checkpoint/restore."""
    config: ContinualConfig
    model_state: Dict[str, Any]
    fisher_snapshots: List[Any]
    buffer_snapshot: List[Any]
    column_registry: List[Any]
    transfer_matrix: Optional[Any]
    retention_matrix: Optional[RetentionMatrix]
    tasks_seen: List[str]

class ContinualOrchestrator:
    """Capstone: unified continual learning pipeline."""

    def __init__(
        self,
        model: nn.Module,
        config: ContinualConfig,
        consolidator: Any = None,   # ElasticWeightConsolidator
        expander: Any = None,       # ProgressiveNetExpander
        replay_mgr: Any = None,     # ReplayMemoryManager
        scheduler: Any = None,      # CurriculumScheduler
    ) -> None:
        self._model = model
        self._config = config
        self._consolidator = consolidator
        self._expander = expander
        self._replay = replay_mgr
        self._scheduler = scheduler
        self._tasks_seen: List[str] = []
        self._retention = RetentionMatrix(task_ids=[], matrix=torch.empty(0, 0))
        self._strategy_log: List[Dict[str, Any]] = []

    # ── public API ──────────────────────────────────────────────
    async def train_task(
        self, task_id: str, train_loader: Any, val_loader: Any, epochs: int = 10
    ) -> ContinualReport:
        """Train on a new task using the configured CL strategy."""
        t0 = time.time()
        strategy = self._select_strategy(task_id) if self._config.strategy == "auto" else self._config.strategy

        # Pre-training evaluation (FWT)
        fwt_acc = self._evaluate(val_loader)

        # Strategy-specific training
        if strategy in ("regularization", "hybrid"):
            self._train_with_ewc(train_loader, epochs)
        if strategy in ("replay", "hybrid"):
            self._train_with_replay(train_loader, epochs)
        if strategy in ("architectural", "hybrid"):
            self._expand_and_train(task_id, train_loader, epochs)
        if strategy == "regularization":
            self._train_with_ewc(train_loader, epochs)

        # Post-training consolidation
        if self._consolidator:
            self._consolidator.compute_fisher(task_id, train_loader)
        if self._replay:
            self._store_replay_data(train_loader, task_id)

        # Update retention matrix
        self._tasks_seen.append(task_id)
        self._update_retention(val_loader, task_id)

        acc = self._evaluate(val_loader)
        elapsed = time.time() - t0

        report = ContinualReport(
            task_id=task_id,
            accuracy=acc,
            avg_accuracy=self._avg_accuracy(),
            backward_transfer=self._retention.bwt(),
            forward_transfer=fwt_acc,
            forgetting_rate=self._forgetting_rate(),
            strategy_used=strategy,
            elapsed_seconds=elapsed,
        )
        self._strategy_log.append({
            "task": task_id, "strategy": strategy, "bwt": report.backward_transfer,
        })
        return report

    def checkpoint(self, path: str) -> None:
        """Save full pipeline state."""
        state = PipelineState(
            config=self._config,
            model_state=self._model.state_dict(),
            fisher_snapshots=[],   # serialise from consolidator
            buffer_snapshot=[],     # serialise from replay mgr
            column_registry=[],     # serialise from expander
            transfer_matrix=None,
            retention_matrix=self._retention,
            tasks_seen=self._tasks_seen,
        )
        torch.save(state, path)

    @classmethod
    def from_checkpoint(cls, path: str, model: nn.Module) -> "ContinualOrchestrator":
        """Restore pipeline from checkpoint."""
        state: PipelineState = torch.load(path)
        model.load_state_dict(state.model_state)
        orch = cls(model=model, config=state.config)
        orch._tasks_seen = state.tasks_seen
        if state.retention_matrix:
            orch._retention = state.retention_matrix
        return orch

    def detect_scenario(self, tasks: List[Any]) -> Literal["task_inc", "domain_inc", "class_inc"]:
        """Classify CL scenario per van de Ven & Tolias (2019)."""
        has_task_labels = all(hasattr(t, "task_label") for t in tasks)
        same_input_dist = len(set(t.domain for t in tasks)) == 1
        if has_task_labels:
            return "task_inc"
        elif same_input_dist:
            return "domain_inc"
        else:
            return "class_inc"

    def retention_report(self) -> Dict[str, Any]:
        """Generate knowledge retention summary."""
        return {
            "tasks_seen": len(self._tasks_seen),
            "avg_accuracy": self._avg_accuracy(),
            "bwt": self._retention.bwt(),
            "fwt": self._retention.fwt(),
            "forgetting_rate": self._forgetting_rate(),
            "retention_matrix_shape": list(self._retention.matrix.shape),
            "strategy_log": self._strategy_log[-10:],
        }

    # ── internals ───────────────────────────────────────────────
    def _select_strategy(self, task_id: str) -> str:
        """Auto-select best strategy based on scenario and history."""
        if not self._tasks_seen:
            return "regularization"
        recent_bwt = self._retention.bwt()
        if recent_bwt < -0.1:
            return "hybrid"     # heavy forgetting → use everything
        elif recent_bwt < -0.03:
            return "replay"     # moderate forgetting → replay helps most
        else:
            return "regularization"  # stable → lightweight EWC sufficient

    def _train_with_ewc(self, loader: Any, epochs: int) -> None: ...
    def _train_with_replay(self, loader: Any, epochs: int) -> None: ...
    def _expand_and_train(self, task_id: str, loader: Any, epochs: int) -> None: ...
    def _store_replay_data(self, loader: Any, task_id: str) -> None: ...
    def _update_retention(self, val_loader: Any, task_id: str) -> None: ...
    def _evaluate(self, loader: Any) -> float: ...
    def _avg_accuracy(self) -> float: ...
    def _forgetting_rate(self) -> float: ...
```

### Algorithm — Adaptive Strategy Selection

```
procedure AUTO_STRATEGY_SELECT(retention_matrix, tasks_seen):
    if |tasks_seen| == 0:
        return "regularization"          // lightweight for first task

    bwt ← compute_backward_transfer(retention_matrix)

    if bwt < -0.10:
        return "hybrid"                  // severe forgetting: all methods
    elif bwt < -0.03:
        return "replay"                  // moderate: replay most effective
    else:
        return "regularization"          // stable: EWC sufficient

procedure COMPUTE_BWT(R, T):
    // R: T×T retention matrix
    sum ← 0
    for i in 1..T-1:
        sum += R[i, T] - R[i, i]        // final accuracy minus peak accuracy
    return sum / (T - 1)

procedure COMPUTE_FWT(R, T, baseline):
    sum ← 0
    for i in 2..T:
        sum += R[i, i-1] - baseline[i]  // zero-shot vs random init
    return sum / (T - 1)
```

### Type Table (mypy-compatible)

| Symbol | Type | Description |
|--------|------|-------------|
| `ContinualConfig.strategy` | `Literal[...]` | Pipeline strategy |
| `ContinualReport.backward_transfer` | `float` | BWT metric |
| `ContinualReport.forward_transfer` | `float` | FWT metric |
| `ContinualReport.forgetting_rate` | `float` | Fraction of degraded tasks |
| `RetentionMatrix.matrix` | `Tensor` | T×T accuracy matrix |
| `PipelineState` | `dataclass` | Full checkpoint state |
| `detect_scenario` return | `Literal["task_inc", "domain_inc", "class_inc"]` | CL scenario |
| `retention_report` return | `Dict[str, Any]` | Summary metrics |

**Comment by web3guru888** [comment]
## 🧠 Phase 34 Launched — Neuromorphic Computing & Spiking Neural Network Integration

Phase 33 (Continual Learning & Catastrophic Forgetting Prevention) is **COMPLETE** ✅. Launching Phase 34!

### Phase 34 Sub-phases

| Sub-phase | Issue | Show & Tell | Q&A | Wiki |
|-----------|-------|-------------|-----|------|
| 34.1 SpikingNeuronModel | #706 | #714 | #717 | [Wiki](https://github.com/web3guru888/asi-build/wiki/Phase-34-Spiking-Neuron-Model) |
| 34.2 EventDrivenProcessor | #707 | #712 | #713 | [Wiki](https://github.com/web3guru888/asi-build/wiki/Phase-34-Event-Driven-Processor) |
| 34.3 SynapticPlasticityEngine | #708 | #715 | #716 | [Wiki](https://github.com/web3guru888/asi-build/wiki/Phase-34-Synaptic-Plasticity-Engine) |
| 34.4 NeuromorphicEncoder | #710 | #718 | #720 | [Wiki](https://github.com/web3guru888/asi-build/wiki/Phase-34-Neuromorphic-Encoder) |
| 34.5 NeuromorphicOrchestrator | #711 | #719 | #721 | [Wiki](https://github.com/web3guru888/asi-build/wiki/Phase-34-Neuromorphic-Orchestrator) |

**Planning discussion**: #709

### Key Components
- **SpikingNeuronModel** — LIF, Izhikevich (20+ patterns), Hodgkin-Huxley, AdEx neuron models with STDP
- **EventDrivenProcessor** — Async event queue, sparse activation, Loihi/TrueNorth/SpiNNaker hardware abstraction
- **SynapticPlasticityEngine** — STDP (asymmetric/symmetric/triplet), R-STDP, homeostatic scaling, structural plasticity
- **NeuromorphicEncoder** — Rate/temporal/population coding, ANN-to-SNN conversion (<2% accuracy loss)
- **NeuromorphicOrchestrator** — Unified pipeline, hybrid SNN-ANN architectures, N-MNIST/DVS-CIFAR10/SHD benchmarks

> *"The third generation of neural network models: networks of spiking neurons."* — Maass (1997)

**Comment by web3guru888** [comment]
## 🚀 Phase 34 Launched — Neuromorphic Computing & Spiking Neural Network Integration

**Phase 33 ✅ COMPLETE** → **Phase 34 🚀**

Building on continual learning (Phase 33), Phase 34 explores brain-inspired neuromorphic computing with event-driven spiking neural networks.

### Phase 34 Components
| Sub-phase | Component | Issue | Focus |
|---|---|---|---|
| 34.1 | SpikingNeuronModel | #706 | HH, Izhikevich, LIF neurons + STDP (Hodgkin & Huxley 1952, Izhikevich 2003) |
| 34.2 | EventDrivenProcessor | #707 | Async event-driven computation (Loihi, TrueNorth emulation) |
| 34.3 | SynapticPlasticityEngine | #708 | STDP, homeostatic plasticity, R-STDP (Bi & Poo 1998) |
| 34.4 | NeuromorphicEncoder | #710 | Rate/temporal/population coding, ANN↔SNN conversion |
| 34.5 | NeuromorphicOrchestrator | #711 | Unified SNN pipeline with energy efficiency metrics |

### Key References
- Hodgkin & Huxley (1952) — Ion-channel neural dynamics
- Izhikevich (2003) — Simple model of spiking neurons
- Bi & Poo (1998) — Spike-timing-dependent plasticity
- Merolla et al. (2014) — IBM TrueNorth
- Davies et al. (2018) — Intel Loihi
- Tavanaei et al. (2019) — Deep learning in SNNs survey

📋 Planning: #709 | 📖 Wiki: 222 pages
